patroni/ha.py

```diff
@@ -128,6 +128,19 @@ def is_leader(self):
         with self._is_leader_lock:
             return self._is_leader > time.time()
 
+    def is_static_primary(self):
+        """Check if this node is configured as the static primary of the cluster."""
+        static_primary = self.patroni.config.get('static_primary')
```

Comment:
Hmm, well the DCS config would be updated on the next loop, and given the nature of the static_primary config (that it dictates cluster leadership) I think it should only be considered once it has been synced to the DCS. Otherwise there is no guarantee that other members will see the config.

Thoughts?

Reply:
> the DCS config would be updated on the next loop

That's exactly my point.
If the value of `static_primary` was changed we need to take action ASAP, not in `loop_wait` seconds.

Reply:
The question I am left with in my mind is this: is it safe? IMHO, I am tempted to say that we should ONLY consider the DCS config to be authoritative.

My working mental model is that, for a configuration element like this, it would really only be safe to take the DCS as authoritative. In such a model, the local config is not considered authoritative for this element.

Consider the case where we have a replica which has been added to a cluster, but the initially given local config did not include the `static_primary` element which is present in the DCS. This could become problematic if we treat the local config as authoritative for this element. Instead, if we treat only the DCS as authoritative, then divergent local configuration is forced to conform to the DCS for this element, ensuring that safety considerations are upheld.

I do believe that this is probably why Patroni is currently implemented (irrespective of this PR) to abort startup if the DCS can not be contacted.

Thoughts? Also, thanks for the discussion and consideration.

Reply:
> we should ONLY consider the DCS config to be authoritative.

Yes, in the ideal world we should, but it is not always possible.
This specific feature could be built only by relying on the DCS config cached by Patroni.

> Consider the case where we have a replica which has been added to a cluster, but the initially given local config did not include the static_primary element

Ok, now I see what you don't understand. There is no such "local" config.
Here is what we have:
1. The global config that is stored in DCS (typically in the /config key). Every heart-beat cycle Patroni loads (at least tries) the cluster view by calling https://github.com/zalando/patroni/blob/fb06af9adbf1267dbc0da72c2d8c3eaf802561a6/patroni/ha.py#L135-L141. As a result, the cluster view is cached in the HA, and in the DCS implementation. The last one could be accessed via the `cluster` property https://github.com/zalando/patroni/blob/fb06af9adbf1267dbc0da72c2d8c3eaf802561a6/patroni/dcs/__init__.py#L763-L766.
2. There is also a cached config. It is stored in the `$PGDATA/patroni.dynamic.json`. After every heart-beat cycle it could be updated from https://github.com/zalando/patroni/blob/fb06af9adbf1267dbc0da72c2d8c3eaf802561a6/patroni/__main__.py#L110-L112
3. Last, there is `bootstrap.dcs`, which only works if DCS is empty and there is no `patroni.dynamic.json`.

When Patroni starts, it first loads config from the `patroni.dynamic.json`, and updates it by whatever is stored in DCS. If both are missing, it will use the `bootstrap.dcs` as the last resort: https://github.com/zalando/patroni/blob/fb06af9adbf1267dbc0da72c2d8c3eaf802561a6/patroni/__main__.py#L38-L49

> I do believe that this is probably why Patroni is currently implemented (irrespective of this PR) to abort startup if the DCS can not be contacted.

Exactly, on start we always want to see what is stored in DCS.

> Thoughts?

Coming back to my original comment. The `self.patroni.config` is updated only _after_ the heart-beat loop. The most up-to-date value is available in the `self.cluster.config`. Hence, the `is_static_primary()`  and `is_static_primary_configured()` methods should be implemented similar to the `get_standby_cluster_config()`.

Reply:
> There is no such "local" config.

Thank you for taking the time to clarify all of that for me. Such being the case, I will go ahead and update this code as you have suggested.

Reply:
Done.